### ISSUE TITLE
[$1000] Reimburse- Error message is displayed after quick input of Rate and Unit

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#15996]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to URL https://staging.new.expensify.com/
2. Log with any account
3. Select any workspace
4. Select tab Reimburse expenses then Track distance
5. Quickly enter the numbers so that 000 does not appear in succession. Immediately select Unit

## Expected Result:
No error message is displayed after quick input of Rate and Unit

## Actual Result:
Error message is displayed after quick input of Rate and Unit

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.2.85.0

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/225386888-596eaa6c-6e2e-46ba-9f2d-266b0d3a3457.mp4

https://user-images.githubusercontent.com/93399543/225386896-b2ef94b5-fda4-465a-b882-a70c0579bff2.mp4

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~011e6f8ab17a2387c5</li>
        <li>Upwork Job ID: 1640484579872690176</li>
        <li>Last Price Increase: 2023-04-11</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @sonialiap (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
Triggered auto assignment to @mateocole (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by sonialiap** [comment]
Going OOO until the 25th, reassigning

**Comment by MelvinBot** [comment]
@mateocole Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by mateocole** [comment]
This looks like a bug, going to mark this as external

**Comment by MelvinBot** [comment]
Triggered auto assignment to @joelbettner (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by mateocole** [comment]
Okay thanks @joelbettner, going to mark as external

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~011e6f8ab17a2387c5

**Comment by MelvinBot** [comment]
Triggered auto assignment to @michaelhaxhiu (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @eVoloshchak (`External`)

**Comment by MelvinBot** [comment]
Current assignee @joelbettner is eligible for the External assigner, not assigning anyone new.

**Comment by podrascanindragan47** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
In Reimburse, the app shows an error message if the user changes the unit soon after changing the rate.

### What is the root cause of that problem?
The app uses two APIs to update the custom unit and rate for a workspace: `UpdateWorkspaceCustomUnitRate` and `UpdateWorkspaceCustomUnit`.
However, an error occurs if the user changes the unit while the app is still updating the rate. This is because the app sends a request to `UpdateWorkspaceCustomUnit` API before the previous request to `UpdateWorkspaceCustomUnitRate` API is completed.

### What changes do you think we should make in order to solve the problem?
To avoid this error, the app should wait for the response from `UpdateWorkspaceCustomUnitRate` API before sending a request to `UpdateWorkspaceCustomUnit` API.
We can use a semaphore to synchronize the two API calls.
The semaphore will ensure that the app only sends a request to UpdateWorkspaceCustomUnit API after receiving a response from UpdateWorkspaceCustomUnitRate API.

**Comment by khatch33** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When quickly changing both the rate and rate unit in the track distance section of the workspace reimbursement page, multiple error messages are displayed: 
-Your changes couldn't be saved. The workspace was modified while you were offline, please try again.
-The policy was modified since you loaded it, please reload the page and try again.

@joelbettner also pointed out that the entire selector has issues (specifically adding 3 decimal places automatically to the end of the rate before typing is done) and should be fixed as a whole.
### What is the root cause of that problem?
The primary problem occurs when the requests with the command UpdateWorkspaceCustomUnitRate and UpdateWorkspaceCustomUnit originating from WorkspaceReimburseView.js happen at nearly the same time. Updating the unit and the unit rate both modify the same policyID so when they both have the same lastModified value in the post request, the second request will respond with an error since the newly updated lastModified value won't match the formData. I was able to reproduce the error for both requests depending on which came second.
<img width="995" alt="Screen Shot 2023-03-27 at 11 12 13 PM" src="https://user-images.githubusercontent.com/95315047/228118583-1ab58e26-1f84-41e6-acd9-0e26fabefb91.png">
<img width="1030" alt="Screen Shot 2023-03-27 at 11 12 33 PM" src="https://user-images.githubusercontent.com/95315047/228118604-708673a9-0983-4163-9b5f-2294f5a51c59.png">
The unit rate change function is debounced so the error will occur almost exclusively when the unit is modified immediately after the unit rate. 
As for the rate changing before the user is done typing, the debounced function is called on every input change. Every input change followed by a full second of no input change will trigger a request and display change.
### What changes do you think we should make in order to solve the problem?
I agree with @joelbettner that a fix should likely go beyond the unit selection bug. However a non-workaround could include debouncing the unit update as well or unit requests being dependent on the unit rate matching the server.

If the unit rates prematurely updating is an issue, fixing it should be done together with the bug listed since it may alter the change needed. Sending requests for both values separately on every input change seems unnecessary and could lead to large accidentally changes from the user. The most intuitive solution would be to combine both input values into one request.

**Comment by ShogunFire** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

1) There is an error message when we select the unit just after editing the rate in workplace track distance screen
2) The app tries to add 3 decimal places automatically after some time and it can be bad ux for the user
3) This form is inconsitent with the other settings form because there is not a save button

### What is the root cause of that problem?

1) The call to UpdateWorkspaceCustomUnitRate and UpdateWorkspaceCustomUnit are almost at the same moment and this can cause problem in the database
2) The Rate is modified on TextChanged with a debounce of 1 second
3) No save button

### What changes do you think we should make in order to solve the problem?

I propose that we only have one call to the API  to modify the rate and the unit and that this method be called when we click on a save button that we need to add.

### What alternative solutions did you explore? (Optional)

**Comment by eVoloshchak** [comment]
I agree with @joelbettner on [this](https://github.com/Expensify/App/issues/15996#issuecomment-1484973151)
> I think the scope of fixing this should be fixing the entire track distance selector and not just the error that we get when switching units

So far we have two separate things that are wrong with selector:

1. Three decimal places are added automatically to the end of the rate before you're done typing
2. Error message is displayed after quick input of Rate and Unit

Are there any other issues?
What should be the correct behavior for the first item? I think we shouldn't update the input unless network request was rejected, do you have any thoughts on this?

**Comment by eVoloshchak** [comment]
This isn't isolated to two different requests, it can also happen with one request being called twice (or more). 
This issue can be reproduced even when you quickly change only the unit, you don't have to change the rate.
It's easier to reproduce with a slow connection


https://user-images.githubusercontent.com/9059945/228575859-3a6bbbf6-b2d6-4b2c-933e-a633a3bd1495.mov

So I don't think the root cause in @podrascanindragan47's [proposal](https://github.com/Expensify/App/issues/15996#issuecomment-1486048909) is entirely correct 

> error occurs if the user changes the unit while the app is still updating the rate. This is because the app sends a request to UpdateWorkspaceCustomUnit API before the previous request to UpdateWorkspaceCustomUnitRate API is completed.

However, I think the proposed solution is pretty close to what is needed 

> To avoid this error, the app should wait for the response from UpdateWorkspaceCustomUnitRate API before sending a request to UpdateWorkspaceCustomUnit API.

But we also need to wait for the response from `UpdateWorkspaceCustomUnitRate` before calling it again and the same with `UpdateWorkspaceCustomUnit`. 
I believe there is a mechanism in the app to queue up requests, correct me if I'm wrong.

**Comment by eVoloshchak** [comment]
@khatch33's [proposal](https://github.com/Expensify/App/issues/15996#issuecomment-1486202310) is more detailed, and explains **why** this error occurs. 
The root cause of this is indeed `lastModified` value.  If you make a second request while the first one is still in progress, the first request will update the report, which will make `lastModified` value in the second request outdated. 

> a non-workaround could include debouncing the unit update as well or unit requests being dependent on the unit rate matching the server.


I don't think debouncing unit requests will fully solve the problem for requests that take more time than the debounce period, but debouncing it in addition to an actual fix seems like a good idea.

> The most intuitive solution would be to combine both input values into one request.

That wouldn't work, since issue is still present even for a single network request

**Comment by eVoloshchak** [comment]
@ShogunFire's [proposal](https://github.com/Expensify/App/issues/15996#issuecomment-1487763981):

> I propose that we only have one call to the API to modify the rate and the unit and that this method be called when we click on a save button that we need to add.

I think not having a `Save` button is a design choice, there are a few screens throughout the app that have 'autosave'

**Comment by khatch33** [comment]
> That wouldn't work, since issue is still present even for a single network request
@eVoloshchak I think combining them into a single write and also increasing the debounced time would mostly solve the issue, but like you said it an error is still possible for a single request particularly with slow connection. I should elaborated more in my proposal but waiting for a response before sending another request is what I was alluding to. However I think even with that, it would be more intuitive to combine both inputs into 1 write especially since they modify the same policyID.  I actually think the best option would be a save or confirm button like @ShogunFire proposed but I figured that was a design choice so I left that out. It seems unnecessary to send requests on every edit and could lead to some pretty big accidents from the user. I also agree debouncing the unit as well doesn't truly fix the underlying issue, it would just make it much harder to reproduce the bug.

**Comment by podrascanindragan47** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/15996#issuecomment-1486048909)

**Comment by mateocole** [comment]
Going to be OOO, reassigning

**Comment by MelvinBot** [comment]
Current assignee @michaelhaxhiu is eligible for the Bug assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by eVoloshchak** [comment]
@khatch33 
> waiting for a response before sending another request is what I was alluding to. However I think even with that, it would be more intuitive to combine both inputs into 1 write especially since they modify the same policyID

We could do that, but that wouldn't solve our problem, it occurs even for a single network request being called multiple times in a row.

> It seems unnecessary to send requests on every edit and could lead to some pretty big accidents from the user. I actually think the best option would be a save or confirm button like @ShogunFire proposed

This is a valid argument, and that would also prevent our issue from happening (we can just disable the save button while another request is in progress). However, not having the save button is a design choice, so I think it's worth starting a discussion on slack (I'll do that shortly).

**Comment by podrascanindragan47** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
In Reimburse, the app shows an error message if the user changes the unit soon after changing the rate.

### What is the root cause of that problem?
The proposal of @khatch33 is correct for this.

### What changes do you think we should make in order to solve the problem?
Following @eVoloshchak ’s comment, I used a queue to handle the API calls.
I created an updateQueue property in the WorkspaceReimburseView constructor and a processUpdateQueue method to execute the queue items one by one.
```jsx
    this.updateQueue = [];
```
```jsx
    processUpdateQueue(func)
    {
        this.updateQueue.push(func);
        if(this.updateQueue.length == 1)
        {
            this.updateQueue[0]();
        }
    }
```
I added the queue processing logic in the componentDidUpdate function.
```
        if(prevProps.policy.lastModified !== this.props.policy.lastModified)
        {
            if(this.updateQueue.length > 0)
            {
                this.updateQueue.shift();
            }

            if(this.updateQueue.length > 0)
            {
                this.updateQueue[0]();
            }
        }
```
I replaced the direct API calls with queue items.
```jsx
        this.processUpdateQueue(() => {
            Policy.updateWorkspaceCustomUnit(this.props.policy.id, distanceCustomUnit, {
                customUnitID: this.state.unitID,
                name: this.state.unitName,
                attributes: {unit: value},
            }, this.props.policy.lastModified);
        });
```
```jsx
        this.processUpdateQueue(() => {
            Policy.updateCustomUnitRate(this.props.policy.id, currentCustomUnitRate, this.state.unitID, {
                ...currentCustomUnitRate,
                rate: numValue * CONST.POLICY.CUSTOM_UNIT_RATE_BASE_OFFSET,
            }, this.props.policy.lastModified);
        });
```

This is result video.
<video src="https://user-images.githubusercontent.com/78560131/229194260-a02b62b7-f8f7-46eb-a4e3-2904de1b1156.mp4" />

**Comment by hellohublot** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Reimburse- Error message is displayed after quick input of Rate and Unit

### What is the root cause of that problem?

After the `first request`, the server is already created the new `lastModified`, the `first request response` has not been returned, the `second request` brings the `old lastModified` to request the api, then the server returns the error


### What changes do you think we should make in order to solve the problem?

https://github.com/Expensify/App/pull/12162

I can reproduce this problem on two different devices. I think if we want to solve it completely, it is best to merge the two interfaces [`UpdateWorkspaceCustomUnit`, `UpdateWorkspaceCustomUnitRate`] into one interface, and we no longer need `lastModified`


### What alternative solutions did you explore? (Optional)
I thought about adding a write lock, if there is a lock, disable all textInput

**Comment by eVoloshchak** [comment]
@podrascanindragan47, thanks for the updated proposal!
We already have a network queue in https://github.com/Expensify/App/tree/main/src/libs/Network that you can make use of. 
However, it seems like this can be reproduced using two devices, which means that queuing up requests won't resolve this completely, so please hold on for now.

**Comment by eVoloshchak** [comment]
> I can reproduce this problem on two different devices

@hellohublot, that is interesting. Could you please provide the exact steps to reproduce this using two devices?

> I think if we want to solve it completely, it is best to merge the two interfaces [UpdateWorkspaceCustomUnit, UpdateWorkspaceCustomUnitRate] into one interface, and we no longer need lastModified

Wouldn't that break the test case from https://github.com/Expensify/App/pull/12162?
If there is no `lastModified` value sent, there could be a case where you:
1. Update unit rate or type while offline on device A
2. Successfully update unit rate or type on a device B while online
3. Bring device A online, the request will be re-tried and will be successful, overriding the changes we did on device B

**Comment by hellohublot** [comment]
@eVoloshchak 

Hello, Thanks. 


> @hellohublot, that is interesting. Could you please provide the exact steps to reproduce this using two devices?
> 
> 

My steps to reproduce are
1. Click `Unit` on device B to pop up `select` in advance, but do not select
2. After inputting `Rate` casually on device A
3. Immediately choose a `Unit` select of device B

The reproduction steps are the `same logic` as the `current` https://github.com/Expensify/App/issues/15996, but there are two devices, which made me give up my original idea of `writing locks`

---


> Wouldn't that break the test case from #12162? If there is no `lastModified` value sent, there could be a case where you:

In fact, when I update the name of the `same workspace`, even if we change the name on offline device A to 'offlinename', and then modify the name on online device B to 'onlinename', after connecting A to the Internet, the workspace name will still be changed to 'offlinename'

If we think this is normal, it also should be normal in this issue

But If we want to solve this problem, it is estimated that updateTime should be carried when the public onyx requests the network (I think lastModified is not suitable than updateTime), 

instead of `adding parameters` to this `single interface`, we adding parameters to `every onyx request`

If the server finds that this is an `offline delay request` and the `updateTime is too early`, then `discard` the request

**Comment by MelvinBot** [comment]
@eVoloshchak @michaelhaxhiu @joelbettner this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by eVoloshchak** [comment]
> My steps to reproduce are
> 
> 1. Click `Unit` on device B to pop up `select` in advance, but do not select
> 2. After inputting `Rate` casually on device A
> 3. Immediately choose a `Unit` select of device B

I can also reproduce this, thank you

> But If we want to solve this problem, it is estimated that updateTime should be carried when the public onyx requests the network (I think lastModified is not suitable than updateTime),
> 
> instead of `adding parameters` to this `single interface`, we adding parameters to `every onyx request`
> 
> If the server finds that this is an `offline delay request` and the `updateTime is too early`, then `discard` the request

This does make sense. 
However, I'm unsure if it's worth it, it would require a lot of changes for something that isn't critical (and kind of hard to reproduce unless you specifically trying to)
@joelbettner, what are your thoughts on [this](https://github.com/Expensify/App/issues/15996#issuecomment-1495260664)?

**Comment by MelvinBot** [comment]
@eVoloshchak, @michaelhaxhiu, @joelbettner Huh... This is 4 days overdue. Who can take care of this?

**Comment by eVoloshchak** [comment]
Not overdue, waiting on opinion from @joelbettner

**Comment by MelvinBot** [comment]
📣 It's been a week! Do we have any satisfactory proposals yet? Do we need to adjust the bounty for this issue? 💸

**Comment by MelvinBot** [comment]
@eVoloshchak @michaelhaxhiu @joelbettner this issue is now 4 weeks old and preventing us from maintaining WAQ, can you:
- Decide whether any proposals currently meet our guidelines and can be approved as-is today
- If no proposals meet that standard, please take this issue internal and treat it as one of your highest priorities
- If you have any questions, don't hesitate to start a discussion in #expensify-open-source

Thanks!

**Comment by michaelhaxhiu** [comment]
Awaiting input from @joelbettner on the value of this fix, and whether or not we should proceed.

**Comment by eVoloshchak** [comment]
Not overdue, waiting on opinion from @joelbettner

**Comment by joelbettner** [comment]
@eVoloshchak the more I think about this, the more I think it is better to have a "Save" button of some sort. Having a "Save" button will limit the amount of unnecessary calls we make to write to the database. For instance...if someone is entering a rate of `123.45`, and they pause after typing `123`, and then pause after typing `.` and then finish by typing out the `45`, that is three seperate calls to that update the databse. That is three seperate commands on the back end that need to be escalated to our leader node.

@michaelhaxhiu do you think it is ok from a design perspective to incorporate a "Save" or "Update" button of some type? That way once the user makes a single change (i.e. changing the distance rate), we make single call to the API and we make a single call to Auth. If this is ok from a UI "looks" perspective, it fits our 1:1:1 call stack design much better.

**Comment by michaelhaxhiu** [comment]
Posted in slack to get more input from the team, as this is a good question for discussion. We'll update here after we get some input.

https://expensify.slack.com/archives/C049HHMV9SM/p1681774424158019

**Comment by michaelhaxhiu** [comment]
From [slack discussion ](https://expensify.slack.com/archives/C049HHMV9SM/p1681828612795059?thread_ts=1681774424.158019&cid=C049HHMV9SM), we are aligned on making a [Save] button it seems!

>We should turn this into a push-to-page form which would have a Save button

>Like the workspace editor would just read as one row, but tapping on the row would take you to a page with two inputs (one for the rate, one for the unit) and a save button

**Comment by eVoloshchak** [comment]
Looks like @ShogunFire [proposal](https://github.com/Expensify/App/issues/15996#issuecomment-1487763981) is what we're looking for, they proposed to do exactly that
> I propose that we only have one call to the API to modify the rate and the unit and that this method be called when we click on a save button that we need to add.

@michaelhaxhiu, do we have a confirmation from the design team on how should this look?

Since we'll be saving both rate and unit simultaneously when Save button is pressed, we don't really need to call two separate endpoints. But we can do it for simplicity's sake since that wouldn't require any changes to the back end. Which one should we do?

**Comment by joelbettner** [comment]
@eVoloshchak

> @michaelhaxhiu, do we have a confirmation from the design team on how should this look?

From the discussion [here](https://expensify.slack.com/archives/C049HHMV9SM/p1681828612795059?thread_ts=1681774424.158019&cid=C049HHMV9SM): 

> Like the workspace editor would just read as one row, but tapping on the row would take you to a page with two inputs (one for the rate, one for the unit) and a save button

> Since we'll be saving both rate and unit simultaneously when Save button is pressed, we don't really need to call two separate endpoints. But we can do it for simplicity's sake since that wouldn't require any changes to the back end. Which one should we do?

I think we should make this a single call to the backend. I can fairly quickly make the changes necessary for that to happen in the backend.

**Comment by eVoloshchak** [comment]
> Looks like @ShogunFire https://github.com/Expensify/App/issues/15996#issuecomment-1487763981 is what we're looking for, they proposed to do exactly that

@ShogunFire already has an active job, new contributors are limited to working on one job at a time.

@joelbettner, does that mean we should open this up for proposals?

**Comment by michaelhaxhiu** [comment]
We can also agree to make an exception here, and allow ShogunFire to take on two jobs? @ShogunFire as long as we ensure you are able to balance the workload and provide updates for both jobs. how do you feel about that?

We've did this recently for another contributor btw.

**Comment by joelbettner** [comment]
@ShogunFire you won't be able to start work on this until the backend changes are complete. Those should be complete and merged soon. I'm currently writing tests for them and they should go into review tomorrow. Just an FYI.

**Comment by ShogunFire** [comment]
> We can also agree to make an exception here, and allow ShogunFire to take on two jobs? @ShogunFire as long as we ensure you are able to balance the workload and provide updates for both jobs. how do you feel about that?

Thanks for this opportunity. I have basically finished the other job. The only problem is that I can't test on all platforms right now. This job sounds more visual so that could be a bigger problem. Also I am not exactly sure what they mean by 
> but tapping on the row would take you to a page with two inputs

Is it showing a modal ? Is there an example in the app of this behaviour ?

**Comment by eVoloshchak** [comment]
> The only problem is that I can't test on all platforms right now. 

[TESTING_MACOS_AND_IOS.md](https://github.com/Expensify/App/blob/main/contributingGuides/TESTING_MACOS_AND_IOS.md) might be of help, please ping me if you have any issues with testing

> Is it showing a modal ? Is there an example in the app of this behaviour ?

I _think_ you can observe a similar UI by creating a workspace and navigating to its general settings


https://user-images.githubusercontent.com/9059945/233607639-bfa2d85f-daf7-4241-8bf0-7689c907e356.mov

**Comment by ShogunFire** [comment]
@eVoloshchak Thank you for the link, I didn't realize it was that cheap.

And thank you for the video. So in Reimburse expenses, instead of this
![IMG_20230421_084538](https://user-images.githubusercontent.com/19537677/233652829-78f8a75e-a649-44c1-83ef-4637074074b9.jpg)

We only have one line: "Track distance >" that opens a new page with the "red content" and a save button. Seems good to me.

**Comment by MelvinBot** [comment]
@eVoloshchak, @michaelhaxhiu, @joelbettner Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by joelbettner** [comment]
@eVoloshchak The backend changes have been deployed to staging and should reach production tomorrow. 

@ShogunFire There is a new API endpoint called `updateWorkspaceCustomUnitAndRate` that should be called rather than individual calls to the `updateWorkspaceCustomUnit` and `updateWorkspaceCustomUnitRate` endpoints. The values that need to be passed to `updateWorkspaceCustomUnitAndRate` the same values that get passed to `updateWorkspaceCustomUnit` and `updateWorkspaceCustomUnitRate`, and they are:
 - `authToken`, 
 - `policyID`, 
 - `customUnit`, 
 - `customUnitRate`, and 
 - `lastModified`.

**Comment by ShogunFire** [comment]
@joelbettner Ok thanks, I will start working on it then, can someone assign me ?

To be able to call the new method of the API I have to put USE_WEB_PROXY to false ?

And just to be sure, this is what we want for the reimburse page right ?
![Track distance](https://user-images.githubusercontent.com/19537677/234452796-7c695e6f-5cd7-438f-a11a-c90ede5cad67.PNG)

**Comment by ShogunFire** [comment]
Hello @eVoloshchak , @michaelhaxhiu   I have started working on this already, can you assign me ? For the reimburse page this are the 2 options I have so far:

1) Inside a Section, it looks good but there is an error because Sections must have titles, I can reproduce the style without using a section though
![Track distance](https://user-images.githubusercontent.com/19537677/234938186-7f110f0c-4df6-447a-8cf0-ae72a40aecec.PNG)
2) Not inside a Section
![Track distance 2](https://user-images.githubusercontent.com/19537677/234938188-6d4bb3ca-942d-4381-ac0b-30a3ca103daa.PNG)

Which one should I use ? 
Also the icon looks weird because it's not green. Should I make a green one, should I ask someone else for a green one or should I remove the icon ? 

For the track distance page I have this:
![Track distance 4](https://user-images.githubusercontent.com/19537677/234938515-187dab69-6c5d-4ffb-8636-a59e6f2048f0.PNG)

The call to the api is already working

**Comment by eVoloshchak** [comment]
@ShogunFire, I think accorging to this
> Like the workspace editor would just read as one row, 

we need to display the rate on reimburse page itself, something like this
<img width="368" alt="Screenshot 2023-05-01 at 11 18 26" src="https://user-images.githubusercontent.com/9059945/235427973-3e7d1fe3-9beb-4df8-9c5d-ad88ee67051c.png">



I'm not sure if we should use the same button as we're using for 'Connect bank account" for consistency

**Comment by ShogunFire** [comment]
Thank @eVoloshchak , it makes sense that we should display the rate. 
Since it's a rate it should probably appear with a "per", for example "0.655 per miles", I don't know if we already have this traduction in the app.

The button Connect bank account is the only button like this in the app I think, because it's a very important feature that can change a lot how the app works. I don't think we should use the same design for track distance. 

I think the right arrow is enough. Like in Personal details:
![Screenshot_2023-05-01-07-21-38-13_4f9154176b47c00da84e32064abf1c48](https://user-images.githubusercontent.com/19537677/235452560-e393ea7a-6997-459a-9a72-5aab74197391.jpg)

**Comment by eVoloshchak** [comment]
@shawnborton, could you please take a look at the updated track distance selector mock-ups?

<img width="377" alt="image" src="https://user-images.githubusercontent.com/9059945/235508405-f77478d0-e130-405b-a4f0-2b95c2b4e519.png">

![image](https://user-images.githubusercontent.com/9059945/235508434-c92e6b80-dbc6-40b1-bfd4-5958e0667cf2.png)

**Comment by eVoloshchak** [comment]
Thanks!
<img width="376" alt="image" src="https://user-images.githubusercontent.com/9059945/235509556-7934ec9b-7550-4dcb-a362-742608d1380c.png">
Apart from the missing padding at the bottom, should we change anything else?

**Comment by shawnborton** [comment]
Sorry, it should stay within that same card style you have that has a title and an illustration. The push-to-page style should just replace this:
<img width="345" alt="image" src="https://user-images.githubusercontent.com/2319350/235510713-7a9a8657-3675-4c58-a458-349c68e4f651.png">

**Comment by eVoloshchak** [comment]
Ah, I see, thank you. This is the updated page
<img width="374" alt="image" src="https://user-images.githubusercontent.com/9059945/235511561-bf241011-e341-4d70-bdca-53a4d1051c6e.png">

**Comment by shawnborton** [comment]
Not quite - the outside edges of the push-to-page row should line up with the text above it. So "Rate" should be left-aligned with the text and title.

**Comment by eVoloshchak** [comment]
Should we also expand the button so there's padding to the left of 'Rate'? 
1
<img width="211" alt="image" src="https://user-images.githubusercontent.com/9059945/235513288-1cea7867-8045-4e5a-8ab5-f3d33e1548c4.png">
or 2?
<img width="223" alt="image" src="https://user-images.githubusercontent.com/9059945/235513194-5f1af291-e24e-45c4-8488-d744b36d0547.png">

**Comment by shawnborton** [comment]
If you can make it touch the outer edges of the card, that would be great. So basically 20px over to the left and right.

**Comment by eVoloshchak** [comment]
<img width="363" alt="image" src="https://user-images.githubusercontent.com/9059945/235514530-ccfce9a7-0f6e-47c6-9234-9fc315b7fe05.png">

The arrow button isn't aligned perfectly on the screenshot, since this is just a prototype, but we'll do that in the PR stage

**Comment by shawnborton** [comment]
Cool, that looks good to me. And to confirm, that's the hover state correct?

**Comment by eVoloshchak** [comment]
Yes, that's a hover state. Thanks for the confirmation and the changes!

**Comment by joelbettner** [comment]
I'm actually not able to assign people to jobs :(

But, @michaelhaxhiu could you assign @ShogunFire?

**Comment by shawnborton** [comment]
Not overdue

**Comment by joelbettner** [comment]
@ShogunFire were you able to be assigned? Any progress on this?

**Comment by eVoloshchak** [comment]
@joelbettner, there's progress in https://github.com/Expensify/App/pull/18454

**Comment by michaelhaxhiu** [comment]
https://github.com/Expensify/App/pull/18454#issuecomment-1551565429

@ShogunFire will merge this to main soon. Let's figure out that bug that @eVoloshchak discovered though because we don't want a regression.

**Comment by joelbettner** [comment]
Discussion happening in the PR. Not overdue.

**Comment by joelbettner** [comment]
The PR for this has been merged. @eVoloshchak @michaelhaxhiu what are our next steps from here?

**Comment by eVoloshchak** [comment]
> @eVoloshchak @michaelhaxhiu what are our next steps from here to officially close this issue?

@joelbettner, there were a couple of regressions found, we're going to fix those in the follow-up PR(s).
After that, we need to wait for this to be deployed to production, wait 7 days and if there are no regressions - complete the BZ checklist and close this one out

**Comment by trjExpensify** [comment]
Should the `reviewing` label be on this issue, a PR has been in review and deployed?

**Comment by ShogunFire** [comment]
Hello @trjExpensify I am not sure how it works yet but since this has been in production for 5 days, doesn't it mean the payment would be in 2 days ? Usually the title becomes Hold For Payment that's why I am asking. Thanks

**Comment by trjExpensify** [comment]
Yeah, 7 days from hitting prod and the issue title _should_ have updated. @michaelhaxhiu is managing this issue, and can confirm on the timeline though!

**Comment by ShogunFire** [comment]
@michaelhaxhiu Can you check that ? I believe the payment should have been done 2 days ago, Thanks

**Comment by eVoloshchak** [comment]
Yeah, the [latest PR](https://github.com/Expensify/App/pull/19748) with regression fixes for this was merged on May 30th, so the payment date should be Jun 7th

This was merged more than 9 business days after the assignment
There were 3 regressions:
$ https://github.com/Expensify/App/issues/19703
$ https://github.com/Expensify/App/issues/19702
$ https://github.com/Expensify/App/issues/19704

As for the BZ checklist, we've ended up creating a new page, so this looks more like a feature request. Since we're using a single endpoint for rate and unit, the original bug is impossible now

**Comment by ShogunFire** [comment]
It looks like @michaelhaxhiu is not here. Can someone else do the the paiement 🙏 ?

**Comment by eVoloshchak** [comment]
@ShogunFire, apologies for the delay
Per [CONTRIBUTING.md](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#payment-for-contributions)
> If you have not received payment after 8 days of the PR being deployed to production and there being no regressions, please email [contributors@expensify.com](mailto:contributors@expensify.com) referencing the GH issue and your GH handle.

Could you do that please? I think that's the most streamlined way of doing this, if you won't receive a response - ping me here

**Comment by michaelhaxhiu** [comment]
Just catching up - sorry I lost track of this because our automation didn't update the GH title with the payment due date.

It sounds like the solution is merged to production and any associated regressions are taken care of? I will review things now to ensure this is true.

**Comment by michaelhaxhiu** [comment]
No need to email contributors@expensify.com, I'm looking!

**Comment by michaelhaxhiu** [comment]
paid.
